accepted/tail-calls.md

```diff
@@ -0,0 +1,1249 @@
+    // shuffling values around within the stack frame. This is similar to the
+    // "parallel-copy problem" in SSA at join points. More details on this
+    // later.
+    //
+    // Additionally, in the case where our current stack argument capacity does
+    // not match the callee's stack argument capacity, we have to save temporary
+    // copies of our caller's frame pointer and return address, and restore them
+    // into the correct stack location after growing or shrinking the stack
+    // argument capacity.
+    //
+    // After the new stack arguments are in place, the rest of the old
+    // stack arguments and old stack slots are now garbage.
+    if const { current_stack_args != callee_stack_args } {
+        // Save the old FP and return address in temporary registers.
+        let old_fp = load FP+0;
+        let ret_addr = load FP+8;
+        // Shuffle the stack arguments into their correct places.
+        compute_args_and_move_them_to_registers_and_stack_slots();
+        // Adjust FP by the stack argument capacity delta, to preserve the exact
+        // capacity required by the callee.
+        FP = FP - const { current_stack_args - callee_stack_args };
+        // Restore the old frame pointer and return address.
+        store FP+8 = ret_addr;
+        store FP+0 = old_fp;
+    } else {
+        compute_args_and_move_them_to_registers_and_stack_slots();
+    }
+
+    //                                                    FP                SP
+    //                                                    |                 |
+    //                                                    V                 V
+    // [ ...caller-frame... | ...new-stack-args... | RA | FP' | ...garbage... ]
+
+    // Deallocate the current stack frame's garbage stack slots.
+    SP = FP;
+
+    //                                                  SP/FP
+    //                                                    |
+    //                                                    V
+    // [ ...caller-frame... | ...new-stack-args... | RA | FP' ]
+
+    // Restore the previous frame's FP.
+    FP = pop;
+
+    //   FP                                          SP
+    //   |                                           |
+    //   V                                           V
+    // [ ...caller-frame... | ...new-stack-args... | RA ]
+
+    // Transfer control to the callee. Do not use a call that pushes a return
+    // address, that is already filled in. Just use a plain unconditional jump.
+    jump $callee;
+}
+```
+
+Under this scheme, regular call paths should be exactly as performant as they
+are today. Tail calls will be as well, if they do not use stack arguments, or
+the tail caller and tail callee use the same number of stack arguments. Tail
+calls where there are different numbers of stack arguments will be penalized,
+however. Potentially fairly egregiously, since copying return addresses around
```

Comment:
from what I know, most call stack branch predictors don't care if you copy the return address around, all they care is that the return branches to where they expect. This is particularly true on ISAs where the return address is stored in the Link Register (nearly every common ISA except x86-32/64), since compilers commonly move that around and move it back when returning.

Reply:
Indeed, in the CPU frontend the machine just sees `call` and `ret` at various RIPs, and doesn't know RSP yet; call pushes on the predictor's internal stack and ret pops (slight simplification wrt misspeculation cleanup but...); the load from the actual in-memory stack will happen in the backend and be verified against the popped (predicted) value but in principle could have been loaded from anywhere.

Reply:
Will remove this concern as well.